**Numerical Methods in Science and Engineering**  
**Chapter:** 3  
**Exercise:** 23  
  
**Author:** Supakorn Suttiruang (Lum) 6031857321  
  

**Problem Statement:** Study the convergence rates for the following system of equations,
$$
\begin{bmatrix}
10 && 7 && 8 && 7 \\
7 && 5 && 6 && 5 \\
8 && 6 && 10 && 9 \\
7 && 5 && 9 && 10
\end{bmatrix}
\begin{Bmatrix}
x_1 \\
x_2 \\
x_3 \\
x_4
\end{Bmatrix}
=
\begin{Bmatrix}
32 \\
23 \\
33 \\
31
\end{Bmatrix}
$$
by using (a) Gauss-Seidel iteration method, (b) the successive over-relaxation method with $\omega = 1.25,1.50,1.75$, and (c) the conjugate gradient method with the inital guess values of $x_1 = x_2 = x_3 = x_4 = 0$. Compare the computed solusions with those obtained from the Gauss elimination method.

**Derivation of Solution:**

**Code Listing and Tabulated Solution:** 

Preamble:

In [1]:
import sympy as sym
import numpy as np

(a) Gauss-Seidel Iteration Method

In [2]:
# Symbolic expression definition
x1, x2, x3, x4 = sym.symbols("x1 x2 x3 x4")

A = sym.Matrix([
    [10, 7, 8, 7],
    [7, 5, 6, 5],
    [8, 6, 10, 9],
    [7, 5, 9, 10]
])

x = sym.Matrix([x1, x2, x3, x4])

b = sym.Matrix([32, 23, 33, 31])

# Obtained system of equations
F = A * sym.Matrix([x1, x2, x3, x4]) - b

In [3]:
# Manipulate equations
fx1, = sym.solve(F[0], x1)
fx2, = sym.solve(F[1], x2)
fx3, = sym.solve(F[2], x3)
fx4, = sym.solve(F[3], x4)

In [4]:
# Initial guesses
x_val = np.asarray([0.0, 0.0, 0.0, 0.0])

# tolerance
tol = 0.001

In [5]:
%%time
# Iteration
iter_count = 0
results = []
while True:
    iter_count += 1
    x_val_old = x_val.copy()
    x_val[0] = fx1.subs([(x1, x_val[0]), (x2, x_val[1]), (x3, x_val[2]), (x4, x_val[3])])
    x_val[1] = fx2.subs([(x1, x_val[0]), (x2, x_val[1]), (x3, x_val[2]), (x4, x_val[3])])
    x_val[2] = fx3.subs([(x1, x_val[0]), (x2, x_val[1]), (x3, x_val[2]), (x4, x_val[3])])
    x_val[3] = fx4.subs([(x1, x_val[0]), (x2, x_val[1]), (x3, x_val[2]), (x4, x_val[3])])
    results.append(
        "{:>10}{:>15.4f}{:>15.4f}{:>15.4f}{:>15.4f}".format(iter_count, x_val[0], x_val[1], x_val[2], x_val[3])
    )
    with np.errstate(divide='ignore'):
        if np.all((np.divide(abs(x_val_old - x_val), x_val) * 100) < tol):
            break

print("{:>10}{:>15}{:>15}{:>15}{:>15}".format("Iteration", "x1", "x2", "x3", "x4"))
if len(results) > 10:
    print("...")
for result in results[-10:]:
    print(result)

 Iteration             x1             x2             x3             x4
...
      1828         1.0020         0.9967         1.0008         0.9995
      1829         1.0020         0.9967         1.0008         0.9995
      1830         1.0020         0.9967         1.0008         0.9995
      1831         1.0020         0.9967         1.0008         0.9995
      1832         1.0020         0.9967         1.0008         0.9995
      1833         1.0020         0.9968         1.0008         0.9995
      1834         1.0020         0.9968         1.0008         0.9995
      1835         1.0020         0.9968         1.0008         0.9995
      1836         1.0020         0.9968         1.0008         0.9995
      1837         1.0019         0.9968         1.0008         0.9995
Wall time: 10.1 s


Gauss-Seidel iteration method took **1837** iterations to converge to 0.001% RPE!

(b) Successive Over-Relaxation Method

In [6]:
# Symbolic expression definition
x1, x2, x3, x4 = sym.symbols("x1 x2 x3 x4")

A = sym.Matrix([
    [10, 7, 8, 7],
    [7, 5, 6, 5],
    [8, 6, 10, 9],
    [7, 5, 9, 10]
])

b = sym.Matrix([32, 23, 33, 31])

# Obtained system of equations
F = A * sym.Matrix([x1, x2, x3, x4]) - b

In [7]:
# Manipulate equations
fx1, = sym.solve(F[0], x1)
fx2, = sym.solve(F[1], x2)
fx3, = sym.solve(F[2], x3)
fx4, = sym.solve(F[3], x4)

For $\omega=1.25$,

In [8]:
# Initial guesses
x_val = np.asarray([0.0, 0.0, 0.0, 0.0])

# tolerance
tol = 0.001

In [9]:
%%time
# Weighting the equations
omega = 1.25

# Iteration
iter_count = 0
results = []
while True:
    iter_count += 1
    x_val_old = x_val.copy()
    x_val[0] = (omega * fx1.subs([(x1, x_val[0]), (x2, x_val[1]), (x3, x_val[2]), (x4, x_val[3])])) + (1 - omega) * x_val_old[0]
    x_val[1] = (omega * fx2.subs([(x1, x_val[0]), (x2, x_val[1]), (x3, x_val[2]), (x4, x_val[3])])) + (1 - omega) * x_val_old[1]
    x_val[2] = (omega * fx3.subs([(x1, x_val[0]), (x2, x_val[1]), (x3, x_val[2]), (x4, x_val[3])])) + (1 - omega) * x_val_old[2]
    x_val[3] = (omega * fx4.subs([(x1, x_val[0]), (x2, x_val[1]), (x3, x_val[2]), (x4, x_val[3])])) + (1 - omega) * x_val_old[3]
    results.append(
        "{:>10}{:>15.4f}{:>15.4f}{:>15.4f}{:>15.4f}".format(iter_count, x_val[0], x_val[1], x_val[2], x_val[3])
    )
    with np.errstate(divide='ignore'):
        if np.all((np.divide(abs(x_val_old - x_val), x_val) * 100) < tol):
            break
            
print("{:>10}{:>15}{:>15}{:>15}{:>15}".format("Iteration", "x1", "x2", "x3", "x4"))
# if len(results) > 10:
#     print("...")
# for result in results[-10:]:
#     print(result)
for result in results:
    print(result)

 Iteration             x1             x2             x3             x4
         1         4.0000        -1.2500         1.0625        -0.0391
         2         3.0654        -0.8469         1.4731         0.0746
         3         2.4363        -0.6047         1.6900         0.2013
         4         2.0540        -0.4798         1.7820         0.3226
         5         1.8421        -0.4299         1.7969         0.4297
         6         1.7428        -0.4249         1.7682         0.5189
         7         1.7138        -0.4439         1.7183         0.5901
         8         1.7253        -0.4734         1.6613         0.6447
         9         1.7575        -0.5051         1.6057         0.6853
        10         1.7972        -0.5341         1.5559         0.7145
        11         1.8369        -0.5581         1.5139         0.7348
        12         1.8723        -0.5763         1.4798         0.7484
        13         1.9015        -0.5888         1.4532         0.7573
      

For $\omega=1.50$,

In [12]:
# Initial guesses
x_val = np.asarray([0.0, 0.0, 0.0, 0.0])

# tolerance
tol = 0.01

In [13]:
%%time
# Weighting the equations
omega = 1.50

# Iteration
iter_count = 0
results = []
while True:
    iter_count += 1
    x_val_old = x_val.copy()
    x_val[0] = (omega * fx1.subs([(x1, x_val[0]), (x2, x_val[1]), (x3, x_val[2]), (x4, x_val[3])])) + (1 - omega) * x_val_old[0]
    x_val[1] = (omega * fx2.subs([(x1, x_val[0]), (x2, x_val[1]), (x3, x_val[2]), (x4, x_val[3])])) + (1 - omega) * x_val_old[1]
    x_val[2] = (omega * fx3.subs([(x1, x_val[0]), (x2, x_val[1]), (x3, x_val[2]), (x4, x_val[3])])) + (1 - omega) * x_val_old[2]
    x_val[3] = (omega * fx4.subs([(x1, x_val[0]), (x2, x_val[1]), (x3, x_val[2]), (x4, x_val[3])])) + (1 - omega) * x_val_old[3]
    results.append(
        "{:>10}{:>15.4f}{:>15.4f}{:>15.4f}{:>15.4f}".format(iter_count, x_val[0], x_val[1], x_val[2], x_val[3])
    )
    with np.errstate(divide='ignore'):
        if np.all((np.divide(abs(x_val_old - x_val), x_val) * 100) < tol):
            break
            
print("{:>10}{:>15}{:>15}{:>15}{:>15}".format("Iteration", "x1", "x2", "x3", "x4"))
if len(results) > 10:
    print("...")
for result in results[-10:]:
    print(result)

 Iteration             x1             x2             x3             x4
...
       603         1.0069         0.9887         1.0027         0.9983
       604         1.0069         0.9888         1.0027         0.9984
       605         1.0068         0.9889         1.0027         0.9984
       606         1.0068         0.9890         1.0027         0.9984
       607         1.0067         0.9891         1.0026         0.9984
       608         1.0066         0.9892         1.0026         0.9984
       609         1.0066         0.9893         1.0026         0.9984
       610         1.0065         0.9894         1.0026         0.9985
       611         1.0064         0.9895         1.0025         0.9985
       612         1.0064         0.9896         1.0025         0.9985
Wall time: 3.92 s


For $\omega=1.75$,

In [14]:
# Initial guesses
x_val = np.asarray([0.0, 0.0, 0.0, 0.0])

# tolerance
tol = 0.01

In [15]:
%%time
# Weighting the equations
omega = 1.75

# Iteration
iter_count = 0
results = []
while True:
    iter_count += 1
    x_val_old = x_val.copy()
    x_val[0] = (omega * fx1.subs([(x1, x_val[0]), (x2, x_val[1]), (x3, x_val[2]), (x4, x_val[3])])) + (1 - omega) * x_val_old[0]
    x_val[1] = (omega * fx2.subs([(x1, x_val[0]), (x2, x_val[1]), (x3, x_val[2]), (x4, x_val[3])])) + (1 - omega) * x_val_old[1]
    x_val[2] = (omega * fx3.subs([(x1, x_val[0]), (x2, x_val[1]), (x3, x_val[2]), (x4, x_val[3])])) + (1 - omega) * x_val_old[2]
    x_val[3] = (omega * fx4.subs([(x1, x_val[0]), (x2, x_val[1]), (x3, x_val[2]), (x4, x_val[3])])) + (1 - omega) * x_val_old[3]
    results.append(
        "{:>10}{:>15.4f}{:>15.4f}{:>15.4f}{:>15.4f}".format(iter_count, x_val[0], x_val[1], x_val[2], x_val[3])
    )
    with np.errstate(divide='ignore'):
        if np.all((np.divide(abs(x_val_old - x_val), x_val) * 100) < tol):
            break
            
print("{:>10}{:>15}{:>15}{:>15}{:>15}".format("Iteration", "x1", "x2", "x3", "x4"))
if len(results) > 10:
    print("...")
for result in results[-10:]:
    print(result)

 Iteration             x1             x2             x3             x4
...
       314         1.0032         0.9949         1.0011         0.9993
       315         1.0031         0.9951         1.0011         0.9993
       316         1.0030         0.9952         1.0011         0.9993
       317         1.0029         0.9953         1.0011         0.9993
       318         1.0029         0.9954         1.0010         0.9994
       319         1.0028         0.9955         1.0010         0.9994
       320         1.0027         0.9956         1.0010         0.9994
       321         1.0027         0.9957         1.0010         0.9994
       322         1.0026         0.9958         1.0009         0.9994
       323         1.0025         0.9959         1.0009         0.9994
Wall time: 1.97 s


It can be seen that as $\omega$ grows larger, the result converges at a lower number of iterations.

(c) Conjugate Gradient Method  

Matrix $[A]$ is likely to be positive definite since its diagonal coefficients are all positive.

In [33]:
# Initial guess value
X = sym.Matrix([0, 0, 0, 0])
R = A * X - b # Residual vector
D = -R # Search direction vector

In [35]:
%%time
# Iteration starts
while True:
    lamb = -(sym.transpose(D) * R)[0]/(sym.transpose(D) * A * D)[0]
    X = X + lamb*D
    R = A * X - b
    error = sym.sqrt((np.transpose(R) * R)[0])
    if error == 0:
        break
    alpha = (sym.transpose(R) * A * D)[0] / (sym.transpose(D) * A * D)[0]
    D = -R + (alpha * D)

display(X.evalf())

Matrix([
[1.0],
[1.0],
[1.0],
[1.0]])

Wall time: 15.6 ms


The conjugate gradient method is significantly faster!

Now we compare the results with the given Guassian Elimination Method MATLAB code,

![alt text](2.png "GELim result")

![alt text](1.png "GELim Wall Time")

The speed of GElim MATLAB code is comparable to that of the conjugate gradient method code implemented here in Python.